rfcs/011-opentracing.md

```diff
@@ -0,0 +1,237 @@
+ - [opentracing-erlang](https://github.com/opentracing-contrib/opentracing-erlang) `otter` version donated to opentracing project.
+ - [original otter](https://github.com/Bluehouse-Technology/otter)
+ - [passage](https://github.com/sile/jaeger_passage)
+ 
+The opentracing philosophy is founded on three pillars:
+- Low overhead: the tracing system should have a negligible performance impact on running services.
+- Application-level transparency: programmers should not need to be aware of the tracing system
+- Scalability
+
+The main addition is to include one of the above mentioned libraries and add instrumentation points into the codebase.
+In initial implementation, there would be a new span started on every HTTP request.
+The following HTTP headers would be used to link tracing span with application specific traces.
+- X-B3-ParentSpanId
+- X-B3-TraceId
+- b3
+
+More information about the use of these headers can be found [here](https://github.com/openzipkin/b3-propagation).
+Open tracing [specification](https://github.com/opentracing/specification/blob/master/specification.md) 
+has a number of [conventions](https://github.com/opentracing/specification/blob/master/semantic_conventions.md) 
+which would be good to follow.
+
+In a nutshell the idea is:
+- Take the reference to Parent span from one of the supported header and pass it to `span_start` call.
+- Construct action name to use in `span_start` call.
+- Call `span_start` from `chttpd:handle_request_int/1`.
+- Pass span in `#httpd{}` record
+- Pass `trace_id` and `parent_span_id` through the stack (extend records if needed)
+- Attach span tags to better identify trace events.
+- Attach span logs at important instrumentation points.
+- Forward spans to external service.
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+- [span](https://github.com/opentracing/specification/blob/1.1/specification.md#the-opentracing-data-model): The "span"
+  is the primary building block of a distributed trace, representing an individual unit of work done in a distributed system.
+  Each Span encapsulates the following state:
+   - An operation name
+   - A start timestamp
+   - A finish timestamp
+   - A set of zero or more key:value `Span Tags`. 
+   - A set of zero or more structured logs (key:value `Span Logs`).
+   - A `SpanContext`
+   - `References` to zero or more causally-related `Spans`
+
+---
+
+# Detailed Description
+
+## Selection of a library
+
+As mentioned earlier, there are two flavours of libraries. None of them is perfect for all use cases.
+The biggest differences in between `otters` and `passage` are:
+
+|                                | otters      | passage                   |
```

Comment:
I found one more difference. The `otters` library (which could be zipkin limitation) doesn't support structured logs. The difference in the API is like follows:
- otters
  ```
  -type info() :: binary() | iolist() | atom() | integer().
  -spec tag(maybe_span(), Text :: info(), Service :: info()) -> maybe_span().
  ```
- passage 
  ```
  -spec log(maybe_span(), Fields, log_options()) -> maybe_span() 
       when
            Fields :: log_fields() | fun (() -> log_fields()).
  ```

Just for this difference alone the `passage` might be a better choice after all. My only worry with passage is its dependencies:
- https://github.com/sile/local.git - A Local Name Registration Facility for Erlang. It is a gen_server based registry which uses public ets table. It is used in order to have unlimited number of named tracers without bottleneck on single registry. 
- https://github.com/sile/passage.git - common utilities shared between multiple tracer types
- https://github.com/sile/thrift_protocol.git - hand crafted thrift encoder/decoder

Overal passage's codebase should be more scalable and it is nicely structured.